core/src/main/scala/org/apache/spark/sql/catalyst/catalog/TiCompositeSessionCatalog.scala

```diff
@@ -0,0 +1,344 @@
+package org.apache.spark.sql.catalyst.catalog
+
+import java.net.URI
+import java.util.concurrent.Callable
+
+import org.apache.spark.sql.{AnalysisException, TiContext}
+import org.apache.spark.sql.catalyst.{FunctionIdentifier, QualifiedTableName, TableIdentifier}
+import org.apache.spark.sql.catalyst.analysis.{EmptyFunctionRegistry, NoSuchDatabaseException}
+import org.apache.spark.sql.catalyst.analysis.FunctionRegistry.FunctionBuilder
+import org.apache.spark.sql.catalyst.catalog.CatalogTypes.TablePartitionSpec
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionInfo}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.types.StructType
+
+/**
+ * Policy of operating composite catalog, with one Ti catalog being either primary or secondary catalog.
+ */
+trait CompositeCatalogPolicy {
+  val primaryCatalog: SessionCatalog
+  val secondaryCatalog: SessionCatalog
+  val tiConcreteCatalog: TiSessionCatalog
+  val legacyCatalog: SessionCatalog
+}
+
+/**
+ * Identical Ti catalog policy.
+ * @param tiContext
+ */
+case class IdentityPolicy(tiContext: TiContext) extends CompositeCatalogPolicy {
+  override val primaryCatalog: SessionCatalog = tiContext.tiConcreteCatalog
+  override val secondaryCatalog: SessionCatalog = tiContext.tiConcreteCatalog
+  override val tiConcreteCatalog: TiSessionCatalog = tiContext.tiConcreteCatalog
+  override val legacyCatalog: SessionCatalog = tiContext.legacyCatalog
+}
+
+/**
+ * Legacy catalog first policy.
+ * @param tiContext
+ */
+case class LegacyFirstPolicy(tiContext: TiContext) extends CompositeCatalogPolicy {
+  override val primaryCatalog: SessionCatalog = tiContext.legacyCatalog
+  override val secondaryCatalog: SessionCatalog = tiContext.tiConcreteCatalog
+  override val tiConcreteCatalog: TiSessionCatalog = tiContext.tiConcreteCatalog
+  override val legacyCatalog: SessionCatalog = tiContext.legacyCatalog
+}
+
+/**
+ * A composition of two catalogs that behaves as a concrete catalog.
+ * @param tiContext
+ */
+class TiCompositeSessionCatalog(tiContext: TiContext)
+    extends SessionCatalog(
+      tiContext.tiConcreteCatalog.externalCatalog,
```

Comment:
How does name conflict resolved?

Reply:
resolved by corresponding policy that configured in sparkConfig.